src/ray/common/source_location.h

```diff
@@ -21,7 +21,13 @@ namespace ray {
 
 // A struct which represents source code location (aka. filename and line
 // number), expected to use only via internal macros.
+//
+// TODO(hjiang): Remove default constructor after we upgrade to C++20, which has
+// designated initializer. Ref:
+// https://en.cppreference.com/w/cpp/language/aggregate_initialization
 struct SourceLocation {
+  SourceLocation() = default;
```

Comment:
not sure if this will work...

what I meant is that you can add an explicit constructor with args, and use the constructor to init, rather than using aggregate.

you can also use an inline function, if you want to keep this a pure struct without any method.

like 

````
inline SourceLocation makeSourceLocation(std::string_view filename, int line_no) {
    SourceLocation loc;
    loc.filename = filename;
    loc.line_no = line_no;
    return loc;
}

#define RAY_LOC() (makeSourceLocation(__FILE__, __LINE__))
````

old school, but works.. and smart compilers often can generate basically the same output code.

Reply:
````
[2024-11-02T16:06:49Z] src/ray/common/source_location_test.cc(35): error C7555: use of designated initializers requires at least '/std:c++20'
[2024-11-02T16:06:49Z] (16:06:49) [31m[1mERROR: [0msrc/ray/common/BUILD:281:12: Compiling src/ray/common/source_location_test.cc failed: (Exit 2): cl.exe failed: error executing command (from target //src/ray/common:source_location_test) 
````

the current form (in this PR) is still failing

the issue is that "designated initializers" is not supported in the current compiler version/flags on windows. just adding a constructor won't help. it needs to avoid using "designated initializers".


----

other than the inline function. what I meant is that you can have a constructor like:

````
struct SourceLocation {
   SourceLocation(std::string_view filename, int line_no): filename(filename), line_no(line_no) {}
   SourceLocation(std::string_view filename): filename(filename), line_no(0) {}
   SourceLocation(const SourceLocation & loc) = default;

   std::string_view filename;
   int line_no;
}

#define RAY_LOC() (SourceLocation(__FILE__, __LINE__))
````

Reply:
Gotcha, fixed.

Reply:
you did not really follow my suggestion..

you would at least need a default copy constructor. otherwise the struct will become not copy-able.

you want to remove the no arg constructor:

````
SourceLocation() = default;
````

Reply:
Default ctor is used elsewhere.

Reply:
> you would at least need a default copy constructor. otherwise the struct will become not copy-able.

Are you sure? C++ compiler by default adds a copy/move ctor and assignment here.

Reply:
> > you would at least need a default copy constructor. otherwise the struct will become not copy-able.
> 
> Are you sure? C++ compiler by default adds a copy/move ctor and assignment here.

no, I am not.. and I was wrong.. yes, the default copy move ctor will be generated unless explicitly deleted.

````
SourceLocation { __FILE__, __LINE__ }
````

this is still using aggregate init rather than using the ctor though, right? it is just not "Designated" now, and this is calling the default no-arg ctor (I think).

if you define a ctor, and want to use it, I think you need to use `SourceLocation ( __FILE__, __LINE__ )`

Reply:
No

Reply:
`{}` is standard and recommanded way to invoke constructor, see effective modern C++ item7

Reply:
> `{}` is standard and recommanded way to invoke constructor, see effective modern C++ item7

yeah.. you are right.. :)

I guess I need to relearn c++..

Reply:
and chatgpt gave me the wrong answer 😅

Reply:
https://chatgpt.com/share/672792a8-97d4-800a-b4e1-2fd3dc8cf08a

Reply:
I guess when constructors are added, it is not called aggregate init anymore..